File: 6. Case-Insensitive Pattern Matching.cpp
// Modify an existing string matching algorithm to perform case-insensitive pattern search.
#include <bits/stdc++.h>
using namespace std;

// Function to convert string to lowercase
string toLowerCase(const string& s)
{
    string res = s;
    for (char& c : res)
    {
        c = tolower(c);
    }
    return res;
}

// Case-insensitive Naive String Matching
void searchPattern(const string& text, const string& pattern)
{
    string txt = toLowerCase(text);
    string pat = toLowerCase(pattern);

    int n = txt.length();
    int m = pat.length();

    for (int i = 0; i <= n - m; i++)
    {
        int j;
        for (j = 0; j < m; j++)
        {
            if (txt[i + j] != pat[j])
                break;
        }
        if (j == m)
            cout << "Pattern found at index " << i << endl;
    }
}

int main()
{
    string text = "Hello World, HELLO Universe!";
    string pattern = "hello";
    searchPattern(text, pattern);
    return 0;
}

/*
Time Complexity (TC): O((N-M+1)*M) where N = length of text, M = length of pattern
Space Complexity (SC): O(N+M) due to lowercase conversion of text and pattern
*/

File: 6. Case-Insensitive Pattern Matching.py
# Modify an existing string matching algorithm to perform case-insensitive pattern search.
def search_pattern(text, pattern):
    txt = text.lower()
    pat = pattern.lower()

    n = len(txt)
    m = len(pat)

    for i in range(n - m + 1):
        j = 0
        while j < m:
            if txt[i + j] != pat[j]:
                break
            j += 1
        if j == m:
            print(f"Pattern found at index {i}")

# Example usage
text = "Hello World, HELLO Universe!"
pattern = "hello"
search_pattern(text, pattern)

"""
Time Complexity (TC): O((N-M+1)*M) where N = length of text, M = length of pattern
Space Complexity (SC): O(N+M) due to lowercase conversion of text and pattern
"""
